tapchannel/aux_closer.go

```diff
@@ -528,38 +529,34 @@ func (a *AuxChanCloser) ShutdownBlob(
 		return none, err
 	}
 
-	// Next, we'll collect all the assets that we own in this channel.
-	assets := commitState.LocalAssets.Val.Outputs
+	// Next, we'll collect all the asset IDs that were committed to the
+	// channel.
+	assetIDs := fn.Map(
+		fundingState.FundedAssets.Val.Outputs,
+		func(o *tapchannelmsg.AssetOutput) asset.ID {
+			return o.AssetID.Val
+		},
+	)
 
 	// Now that we have all the asset IDs, we'll query for a new key for
 	// each of them which we'll use as both the internal key and the script
 	// key.
 	scriptKeys := make(tapchannelmsg.ScriptKeyMap)
-	for idx := range assets {
-		channelAsset := assets[idx]
-
+	for _, assetID := range assetIDs {
 		newKey, err := a.cfg.AddrBook.NextScriptKey(
 			ctx, asset.TaprootAssetsKeyFamily,
 		)
 		if err != nil {
 			return none, err
 		}
 
-		// We now add the a
-		// TODO(guggero): This only works if there's only a single asset
-		// in the channel. We need to extend this to support multiple
-		// assets.
-		_, err = a.cfg.AddrBook.NewAddressWithKeys(
-			ctx, address.V1, channelAsset.AssetID.Val,
-			channelAsset.Amount.Val, newKey, newInternalKey, nil,
-			*a.cfg.DefaultCourierAddr,
-		)
+		err = a.cfg.AddrBook.InsertScriptKey(ctx, newKey, true)
 		if err != nil {
-			return none, fmt.Errorf("error adding new address: %w",
-				err)
+			return none, fmt.Errorf("error declaring script key: "+
+				"%w", err)
 		}
 
-		scriptKeys[channelAsset.AssetID.Val] = *newKey.PubKey
+		scriptKeys[assetID] = *newKey.PubKey
```

Comment:
Based on the comment above, we may end up with script keys thgat was created/inserted but were never used as we overrode them here in the map.

Reply:
Yeah, you're right... So perhaps for the very first one we should try to re-use the one we already derived?

Reply:
Okay, my comment right above this was incorrect. Because we shouldn't have duplicate asset IDs (as explained in https://github.com/lightninglabs/taproot-assets/pull/1413#discussion_r2023105021), we shouldn't overwrite keys here. So we don't derive keys that we then don't end up using.